add language support for arbitrary languages

AmethystGear opened the pull request:
Allows boba support for arbitrary languages.
This is achieved by allowing a new parameter to the `compile` and `run` commands: `--supported_langs` (abbreviated `-sl`).

this parameter accepts a path to a file with valid json containing entries of the format:
```
{
    "c" : {
        "ext" : ["c"],
        "compile" : ["gcc", "-o", "{{universe_name}}", "{{script_name}}"],
        "run" : ["./{{universe_name}}"]
    }
    "python" : {
        "ext": ["py"]
        "run" : ["python", "{{script_name}}"]
    }
.
.
.
}
```
when executing, `{{universe_name}}` is replaced with the name of the universe (`universe_1`, `universe_2`, etc.) and `{{script_name}}` is replaced with the name of the script (`universe_1.c/py/R..` etc.)
which defines languages that are understood by the `Lang` class. 
by default, python and R are supported (see `DEFAULT_LANGS` in `lang.py`)

As shown above, this works for both compiled and interpreted languages.

This PR also contains a simple example for the C language. Simply do
```
boba compile -s template.c --supported_langs lang.json
boba run --all --supported_langs lang.json
```
in the `examples/example_c` folder, and you should see the output:
```
[universe_1] hello from universe 1
[universe_2] hello from universe 2
[universe_3] hello from universe 3
```

AmethystGear committed:
added language support for arbitrary languages

AmethystGear committed:
changed 'compiled' to 'universe_name'

AmethystGear committed:
fixed supported_langs None bug

yyyliu commented on boba/cli.py:
```diff
@@ -16,15 +17,23 @@
-def compile(script, out, lang):
+@click.option('--supported_langs', '-sl', 'supp_langs', help='json file with supported languages',
+              default=None)
+def compile(script, out, lang, supp_langs):
```
Supplying `--supported_langs` every time when the user compiles and runs a script is tedious.

Is it possible to write the path of the json file in the `BOBA_CONFIG` block? Something similar to the visualizer config here: https://github.com/uwdata/boba/blob/8e009dc6298b094c43f3670f5524350399b92cde/example/hurricane/repro.R#L41

yyyliu commented on boba/cli.py:
```diff
@@ -16,15 +17,23 @@
     click.echo('Creating multiverse from {}'.format(script))
-    ps = Parser(script, out, lang)
+    
+    supported_langs = None
+    if supp_langs:
+        with open(supp_langs, 'r') as l:
+            supported_langs = json.load(l)
```
These lines should happen in the parser class ... Here we might simply check if the file exists (using `check_path`).

But if you modify the API to use `BOBA_CONFIG`, we won't modify this file ..

yyyliu commented on boba/lang.py:
```diff
@@ -4,57 +4,65 @@
+        'run' : ['python', script]
+    },
+    'r' : {
+        'ext' : ['R', 'r', 'rmd'],
```
we can't really use `Rscript` to run rmd if it contains custom syntax like \`\`\`{r}```

yyyliu commented on boba/lang.py:
```diff
@@ -4,57 +4,65 @@
-                                'extension {}'.format(ext))
+    def __init__(self, script, lang=None, supported_langs=None):
+        if supported_langs:
+            self.supported_langs = supported_langs
```
should be a combination of supported lang and default lang ... specifically, supported langs can override default lang, but otherwise anything in default lang is still available

yyyliu commented on example/example_c/lang.json:
```diff
@@ -0,0 +1,7 @@
+        "compile" : ["gcc", "-o", "{{universe_name}}", "{{script_name}}"],
+        "run" : ["./{{universe_name}}"]
+    }
+}
```
missing new line

yyyliu commented on example/example_c/template.c:
```diff
@@ -0,0 +1,7 @@
+#include <stdio.h>
```
since the example is so short, I think it should belong to a test case (e.g. using a test case with such input, but do not put it in the example folder)

yyyliu reviewed (changes requested):
Thanks Shiven! My comments are inline.

AmethystGear committed:
fixed PR feedback, added test case for language functionality

AmethystGear committed:
fixed bug in unit test

AmethystGear commented:
fixed the PR feedback and converted the C example to a unit test. Additionally, as requested, the `supported_langs` parameter was removed and replaced with a `lang` variable in the `BOBA_CONFIG`, like so:
```
# --- (BOBA_CONFIG)
{
  "lang": "lang.json"
}
# --- (END)
```
as seen in the c test case (can be found in `test/test_c`)

yyyliu commented:
Thanks!
